src/coreclr/vm/comutilnative.cpp

```diff
@@ -87,111 +123,83 @@ FCIMPL3(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectU
     }
-
-    if (gc.dynamicMethodsArray != NULL)
+    else if (gc.stackTrace.Get() != NULL)
     {
-        // Get the number of elements in the dynamic methods array
-        unsigned   cOrigDynamic = gc.dynamicMethodsArray->GetNumComponents();
-
-        // ..and allocate a new array. This can trigger GC or throw under OOM.
-        gc.dynamicMethodsArrayCopy = (PTRARRAYREF)AllocateObjectArray(cOrigDynamic, g_pObjectClass);
-
-        // Deepcopy references to the new array we just allocated
-        memmoveGCRefs(gc.dynamicMethodsArrayCopy->GetDataPtr(), gc.dynamicMethodsArray->GetDataPtr(),
-                                                  cOrigDynamic * sizeof(Object *));
-
-        fHaveDynamicMethodArray = true;
+        *pStackTraceUnsafe = OBJECTREFToObject(gc.stackTraceCopy.Get());
+    }
+    else
+    {
+        *pStackTraceUnsafe = NULL;
     }
-
-    // Prep to return
-    *pStackTraceUnsafe = fHaveStackTrace?OBJECTREFToObject(gc.stackTraceCopy.Get()):NULL;
-    *pDynamicMethodsUnsafe = fHaveDynamicMethodArray?OBJECTREFToObject(gc.dynamicMethodsArrayCopy):NULL;
-
     HELPER_METHOD_FRAME_END();
 }
 FCIMPLEND
 
 // Given an exception object and deep copied instances of a stacktrace and/or dynamic method array, this method will set the latter in the exception object instance.
-FCIMPL3(VOID, ExceptionNative::SaveStackTracesFromDeepCopy, Object* pExceptionObjectUnsafe, Object *pStackTraceUnsafe, Object *pDynamicMethodsUnsafe);
+FCIMPL2(VOID, ExceptionNative::SaveStackTracesFromDeepCopy, Object* pExceptionObjectUnsafe, Object *pStackTraceUnsafe);
 {
     CONTRACTL
     {
         FCALL_CHECK;
     }
     CONTRACTL_END;
-
+    
     ASSERT(pExceptionObjectUnsafe != NULL);
 
     struct
     {
         StackTraceArray stackTrace;
-        EXCEPTIONREF refException;
-        PTRARRAYREF dynamicMethodsArray; // Object array of Managed Resolvers
+        StackTraceArray stackTraceCopy;
+        OBJECTREF refStackTrace = NULL;
+        PTRARRAYREF refKeepaliveArray = NULL;
+        PTRARRAYREF refKeepaliveArrayCopy = NULL;
+        EXCEPTIONREF refException = NULL;
     } gc;
-    gc.refException = NULL;
-    gc.dynamicMethodsArray = NULL;
 
     // GC protect the array reference
     HELPER_METHOD_FRAME_BEGIN_PROTECT(gc);
 
     // Get the exception object reference
     gc.refException = (EXCEPTIONREF)(ObjectToOBJECTREF(pExceptionObjectUnsafe));
+    gc.refStackTrace = (OBJECTREF)(ObjectToOBJECTREF(pStackTraceUnsafe));
 
-    if (pStackTraceUnsafe != NULL)
-    {
-        // Copy the stacktrace
-        StackTraceArray stackTraceArray((I1ARRAYREF)ObjectToOBJECTREF(pStackTraceUnsafe));
-        gc.stackTrace.Swap(stackTraceArray);
-    }
-
-    gc.dynamicMethodsArray = NULL;
-    if (pDynamicMethodsUnsafe != NULL)
-    {
-        gc.dynamicMethodsArray = (PTRARRAYREF)ObjectToOBJECTREF(pDynamicMethodsUnsafe);
-    }
+    ExceptionObject::GetStackTraceParts(gc.refStackTrace, gc.stackTrace, &gc.refKeepaliveArray);
```

Comment:
This does way too many copies. This will be often called by async on different thread than the exception was thrown originally that will cause `ExceptionObject::GetStackTraceParts` to create a copy. The `DeepCopyStackTrace` is then going to create second copy.

I think we can avoid all copies in this method by:
- Setting the owning thread in `ExceptionNative::GetStackTracesDeepCopy` to (Thread*)1. It will signal that the stack trace is immutable. Any thread that will want to append anything to it will have make its own copy first.
- Once we do that, I think this method can be deleted and restoring of the stacktrace can be replace by a simple assignment, without any copies.

Reply:
I've implemented this suggestion.